MultigridProjector/Logic/MultigridProjection.cs

```diff
@@ -1556,6 +1561,26 @@ private void FindProjectedBlock(Vector3D center, Vector3D reachFarPoint, ref MyW
                 cubes[i] = null;
         }
 
+        public static bool CheckVoxels(MySlimBlock block)
+        {
+            if (MyPerGameSettings.Destruction && block.CubeGrid.GridSizeEnum == MyCubeSize.Large)
+                return block.CubeGrid.Physics.Shape.BlocksConnectedToWorld.Contains(block.Position);
+
+            block.GetWorldBoundingBox(out BoundingBoxD worldBoundingBox);
+            List<MyVoxelBase> voxels = new List<MyVoxelBase>();
+            MyGamePruningStructure.GetAllVoxelMapsInBox(ref worldBoundingBox, voxels);
```

Comment:
Performance: `CanBuild` tends to be called a lot, which means a lot of potential calls to `CheckVoxels` as well, including on server side.

We are getting the list of voxels per invocation (per block), which is wasteful. This list can be cached very effectively.

- Cache the list of voxels inside the bounding box of the **grid** the block is on (not the block).
- `CheckVoxels` can return `false` right away if the the list of voxels is empty.
- The cache is valid as long as the bounding box of the grid does not change.
- The cache should be cleared from entries not used for more than 10 minutes.
- The cache should be invalidated when voxels are added/removed from the world (asteroid spawn).